### ISSUE TITLE
Being able to compute PublicKey from an ExpandedSecretKey

### ISSUE THREAD

**Issue by coltfred** [dalek-cryptography/ed25519-dalek#39]
I'm in a system where the user has an ExpandedSecretKey and I want to sign a message. I don't have access to the SecretKey byes that were used to produce the ExpandedSecretKey. Is there a way to compute the PublicKey without manually multiplying times the base point? Also I'm confused about why the `sign` function takes the PublicKey when it can be computed.

Sorry if I've missed something.

**Comment by ValarDragon** [comment]
Ed25519 signing requires the public key. While it can be computed, that is an extra constant time scalar multiplication, which you could avoid. I believe the choice of having sign take in the pubkey was in acknowledgement of performance. Most of the time you would already have the pubkey before signing, thus theres little sense in recomputing it.

**Comment by coltfred** [comment]
@ValarDragon In the case of the ExpandedSecretKey it's ugly to compute the key because the scalar is hidden as private to the crate and there isn't a `compute_public_key` on it. I don't care if I need to pass it, but having to manually multiply the base point, serialize to bytes and calling Scalar::from_bits is a bit much.

For clarity, here is my code:

```rust
        let private_key_bytes: [u8; 64] = //Some real value
        //This unwrap cannot fail. The only thing that the `from_bytes` does for validation is that the
        //value is 64 bytes long, which we guarentee statically.
        let secret_key: ExpandedSecretKey =
            ExpandedSecretKey::from_bytes(&private_key_bytes).unwrap();
        let secret_bytes: [u8; 32] = {
            let mut secret_bytes = [0; 32];
            secret_bytes.copy_from_slice(&private_key_bytes[0..32]);
            secret_bytes
        };
        let public_key_point =
            (&Scalar::from_bits(secret_bytes) * &constants::ED25519_BASEPOINT_TABLE).compress();
        let sig = secret_key.sign::<Sha512>(
            &message,
            &PublicKey::from_bytes(&public_key_point.to_bytes()).unwrap(),
        );
```

**Comment by coltfred** [comment]
@isislovecruft @ValarDragon  Did you have any thoughts on this? I'd be happy to PR this change if you'd be willing to accept it. I'm thinking adding this to `ExpandedSecretKey`:
```rust
fn compute_public_key(&self) -> PublicKey 
```

**Comment by isislovecruft** [comment]
Hi! This is a great idea. I'm sorry though, yesterday I saw your suggestion and my tab didn't autorefresh for some reason, so I only saw the original comment. :( So.. I've already implemented it in #41. Sorry about that; I didn't mean to step on your toes and would have totally taken a PR.

**Comment by coltfred** [comment]
@isislovecruft No worries. I'm not hurt. I did open a PR with a couple other suggestions, though they could be implemented with `From` like you did for the PublicKey. I thought the `new` pattern was more standard, but I'd love to hear your feedback.

**Comment by coltfred** [comment]
Fixed by #41
